bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/scheduler/ExpressionThreadPoolManager.java

```diff
@@ -195,97 +213,130 @@ public void run() {
+                                                    .schedule(task, delay, TimeUnit.MILLISECONDS);
                                             taskFutures.add(newFuture);
                                             logger.trace("Task '{}' has now {} Futures", task, taskFutures.size());
                                             timestamps.put(newFuture, time);
                                         }
+                                        if (logger.isTraceEnabled()) {
+                                            for (ScheduledFuture<?> future : taskFutures) {
+                                                logger.trace("Task {} ({}) will run in {}", task,
+                                                        System.identityHashCode(task),
+                                                        future.getDelay(TimeUnit.MILLISECONDS));
+                                            }
+                                        }
+                                    } finally {
+                                        futuresLock.unlock();
                                     }
                                 } else {
                                     logger.trace("Expressions without tasks are not valid");
                                 }
 
-                                if (firstExecution == null) {
-                                    firstExecution = time;
+                                if (earliestExecution == null) {
+                                    earliestExecution = time;
                                 } else {
-                                    if (time.before(firstExecution)) {
-                                        firstExecution = time;
+                                    if (time.before(earliestExecution)) {
+                                        earliestExecution = time;
                                     }
                                 }
 
                             } else {
-                                logger.info("Expression '{}' has no future executions anymore", e);
+                                logger.debug("Expression '{}' has no future executions anymore", e);
                                 finishedExpressions.add(e);
                             }
                         }
 
                         for (Expression e : finishedExpressions) {
                             scheduled.remove(e);
+                            logger.trace("Cleaning up finished expression '{}'", e);
                         }
 
-                        if (firstExecution != null) {
-                            while (now.before(firstExecution)) {
+                        if (earliestExecution != null) {
+                            boolean newExpression = false;
+                            while (!newExpression && new Date().before(earliestExecution)) {
+                                long sleepDuration = earliestExecution.getTime() - now.getTime();
```

Comment:
In the line above you check against an "updated now" - isn't the risk here that you slightly sleep too long if you use the "previous now" here?